buildSrc/src/main/kotlin/org/gradle/plugins/pegdown/PegDown.kt

```diff
@@ -0,0 +1,33 @@
+package org.gradle.plugins.pegdown
+
+import org.gradle.api.DefaultTask
+import org.gradle.api.tasks.*
+import org.pegdown.PegDownProcessor
+import java.io.File
+import java.nio.charset.Charset
+
+@CacheableTask
+class PegDown : DefaultTask() {
+    @Input
+    val inputEncoding = Charset.defaultCharset()
+
+    @Input
+    val outputEncoding = Charset.defaultCharset()
+
+    @PathSensitive(PathSensitivity.NONE)
+    @InputFile
+    lateinit var markdownFile: File
```

Comment:
Is this the proposed/accepted way of handling this in tasks? If someone fails to specify this when declaring the task, is the expected behaviour to receive an NPE because of an uninitialized field?

I understand that this isn't a public task, but I'm wondering what the proposed best practices are around fields like this that are initialized by the API consumer. @bamboo?

Reply:
Those are good questions.

Ideally we wouldn't have uninitialized fields at all. That will become easier to achieve as the `Provider/Property` API matures.

In the meantime there are two main choices for required task values, `lateinit` and nullable properties.

> is the expected behaviour to receive an NPE because of an uninitialized field?

From [the Kotlin language reference](https://kotlinlang.org/docs/reference/properties.html#late-initialized-properties-and-variables):
> Accessing a lateinit property before it has been initialized throws a special exception that clearly identifies the property being accessed and the fact that it hasn't been initialized.

For the `markdownFile` property declared above the error message would be something like` lateinit property markdownFile has not been initialized` which is reasonable for an internal task.

For better error reporting and type safety, `nullable` properties work best:

```kotlin
...    
    @Input
    var markdownFile: File? = null

    @TaskAction
    fun process() {
        val theMarkdownFile = markdownFile
            ?: throw IllegalStateException("`markdownFile` property must be set!")
...
    }
```

Reply:
Thanks for the well thought out and really informative answer.

Would you like me to try to capture this in one of the `getting started` documents in the Kotlin DSL repo?